cudax/include/cuda/experimental/__container/vector.cuh

```diff
@@ -0,0 +1,1585 @@
+  //! @note Neither frees not allocates memory if `__first == __last`.
+  void assign(const size_type __count, const _Tp& __value)
+  {
+    if (capacity() < __count)
+    {
+      __buffer __new_buf{__borrow_resource(), __count};
+      this->__fill_n(__new_buf.begin(), __count, __value);
+      __buf_.__swap_allocations(__new_buf);
+    }
+    else
+    {
+      this->__fill_n(__unwrapped_begin(), __count, __value);
+    }
+    __size_ = __count;
+  }
+
+  //! @brief Replaces the content of the vector with the sequence `[__first, __last)`
+  //! @param __first Iterator to the first element of the input sequence.
+  //! @param __last Iterator after the last element of the input sequence.
+  //! @note Neither frees not allocates memory if `__first == __last`. May allocate multiple times in case of input
+  //! iterators.
+  _LIBCUDACXX_TEMPLATE(class _Iter)
+  _LIBCUDACXX_REQUIRES(_CUDA_VSTD::__is_cpp17_input_iterator<_Iter>::value _LIBCUDACXX_AND(
+    !_CUDA_VSTD::__is_cpp17_forward_iterator<_Iter>::value))
+  void assign(_Iter __first, _Iter __last)
+  {
+    __size_ = 0;
+    for (; __first != __last; ++__first)
+    {
+      emplace_back(*__first);
+    }
+  }
+
+  //! @brief Replaces the content of the vector with the sequence `[__first, __last)`
+  //! @param __first Iterator to the first element of the input sequence.
+  //! @param __last Iterator after the last element of the input sequence.
+  //! @note Neither frees not allocates memory if `__first == __last`.
+  _LIBCUDACXX_TEMPLATE(class _Iter)
+  _LIBCUDACXX_REQUIRES(_CUDA_VSTD::__is_cpp17_forward_iterator<_Iter>::value)
+  void assign(_Iter __first, _Iter __last)
+  {
+    const auto __count = static_cast<size_type>(_CUDA_VSTD::distance(__first, __last));
+    if (capacity() < __count)
+    {
+      __buffer __new_buf{__borrow_resource(), __count};
+      this->__copy_cross<_Iter, __detect_transfer_kind<__is_host_only, _Iter>>(
+        __first, __last, __new_buf.begin(), __count);
+      __buf_.__swap_allocations(__new_buf);
+    }
+    else
+    {
+      this->__copy_cross<_Iter, __detect_transfer_kind<__is_host_only, _Iter>>(
+        __first, __last, __unwrapped_begin(), __count);
+    }
+    __size_ = __count;
+  }
+
+  //! @brief Replaces the content of the vector with the initializer_list \p __ilist
+  //! @param __ilist The initializer_list to be copied into this vector.
+  //! @note Neither frees not allocates memory if `__ilist.size() == 0`.
+  void assign(_CUDA_VSTD::initializer_list<_Tp> __ilist)
+  {
+    const auto __count = static_cast<size_type>(__ilist.size());
+    if (capacity() < __count)
+    {
+      __buffer __new_buf{__borrow_resource(), __count};
+      this->__copy_cross(__ilist.begin(), __ilist.end(), __new_buf.begin(), __count);
+      __buf_.__swap_allocations(__new_buf);
+    }
+    else
+    {
+      this->__copy_cross(__ilist.begin(), __ilist.end(), __unwrapped_begin(), __count);
+    }
+    __size_ = __count;
```

Comment:
this pattern repeats again and again in different member functions. why not implement them all by dispatching to one `__assign` implementation?